src/OpenTelemetry.Extensions/Internal/RateLimiter.cs

```diff
@@ -0,0 +1,46 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+using System.Diagnostics;
+
+namespace OpenTelemetry.Extensions.Internal;
+
+internal sealed class RateLimiter
+{
+    private readonly Stopwatch stopwatch = Stopwatch.StartNew();
+    private readonly double creditsPerNanosecond;
+    private readonly long maxBalance; // max balance in ticks
+    private long currentBalance; // last op ticks less remaining balance, using long directly with Interlocked for thread safety
+
+    public RateLimiter(double creditsPerSecond, double maxBalance)
+    {
+        this.creditsPerNanosecond = creditsPerSecond / Stopwatch.Frequency;
```

Comment:
Should this be simply dividing it by 10^9?. Looking at java implementation --> https://github.com/open-telemetry/opentelemetry-java/blob/ad120a5bff0887dffedb9c73af8e8e0aeb63659a/sdk/common/src/main/java/io/opentelemetry/sdk/internal/RateLimiter.java#L36

Reply:
The .NET stopwatch class does not have a fixed frequency for its ticks - it depends on the hardware/OS that it is running on, and provides the `Frequency` property to indicate how many ticks there are a second.
https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.stopwatch.elapsedticks?view=net-8.0#remarks

Reply:
got it- it's the name `creditsPerNanosecond` that confused me. Cannot think of a better name atm.